resim/utils/tuple_utils_test.cc

```diff
@@ -0,0 +1,131 @@
+// Copyright 2023 ReSim, Inc.
+//
+// Use of this source code is governed by an MIT-style
+// license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+#include "resim/utils/tuple_utils.hh"
+
+#include <gtest/gtest.h>
+
+namespace resim {
+
+TEST(TupleUtilsTest, TestForEachInTuple) {
+  // SETUP
+  // NOLINTBEGIN(readability-magic-numbers)
+  auto simple_tuple = std::make_tuple(1, 2., 3U);
+  // NOLINTEND(readability-magic-numbers)
+  const auto add_one = [](auto number) { return number + 1; };
+
+  // ACTION
+  auto result = for_each_in_tuple(add_one, simple_tuple);
+
+  // VERIFICATION
+  static_assert(
+      std::tuple_size_v<decltype(simple_tuple)> ==
+      std::tuple_size_v<decltype(result)>);
+
+  EXPECT_EQ(std::get<0>(simple_tuple) + 1, std::get<0>(result));
+  EXPECT_EQ(std::get<1>(simple_tuple) + 1, std::get<1>(result));
+  EXPECT_EQ(std::get<2>(simple_tuple) + 1, std::get<2>(result));
+}
+
+TEST(TupleUtilsTest, TestForEachInTupleReferences) {
+  // SETUP
+  int a = 1;
+  int b = 2;
+  int c = 3;
+
+  // We can't copy tuples containing rvalue references, so we have to manually
+  // re-construct.
+  using TestTupleType =
+      std::tuple<int, int &, const int &, int &&, const int &&>;
+  // NOLINTBEGIN(readability-magic-numbers)
+  TestTupleType test_tuple{a, b, c, 4, 5};
+  TestTupleType test_tuple_to_move{a, b, c, 4, 5};
+  const TestTupleType test_tuple_const{a, b, c, 4, 5};
+  const TestTupleType test_tuple_const_to_move{a, b, c, 4, 5};
+  // NOLINTEND(readability-magic-numbers)
+
+  const auto identity = [](auto &&x) -> decltype(auto) {
+    return std::forward<decltype(x)>(x);
+  };
+
+  // ACTION
+  auto result = for_each_in_tuple(identity, test_tuple);
+  auto result_moved =
+      for_each_in_tuple(identity, std::move(test_tuple_to_move));
+  auto result_from_const = for_each_in_tuple(identity, test_tuple_const);
+
+  // This one is a bit weird and you wouldn't really ever do this, but we'll
+  // test it anyway.
+  auto result_moved_from_const = for_each_in_tuple(
+      identity,
+      static_cast<const TestTupleType &&>(test_tuple_const_to_move));
```

Comment:
f**king c++

Reply:
It's equivalent to a move, but I want this to be a const rvalue reference. You wouldn't ever do this because you can't "move" or "steal" resources from something that's const, so const T && is strictly less powerful than const T&.. This is mainly included for the sake of completeness.